gateway-spi/src/main/java/org/apache/knox/gateway/topology/Topology.java

```diff
@@ -153,4 +161,106 @@ public void addProvider( Provider provider ) {
     nameMap.put( provider.getName(), provider );
   }
 
+  @Override
+  public int hashCode() {
+    return (new HashCodeBuilder(17, 31)).append(name)
+                                        .append(providerList)
+                                        .append(services)
+                                        .append(applications)
+                                        .build();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```

Comment:
If `Provider`, `Service` and `Application` implements `hashCode` and `equals` properly you do not need the private methods below and you can simply use `EqualsBuilder` here too.

Reply:
The reason for not using EqualsBuilder here is that equals() for Lists requires the ordering to be the same. However, for the purpose of evaluating topologies for equality, such ordering is not important.

Reply:
Not using EqualsBuilder here because it does not handle potentially-null fields well.